General convolution_backward function

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69584
* __->__ #69044
* #68899

Differential Revision: [D32708818](https://our.internmc.facebook.com/intern/diff/D32708818)

jbschlosser committed:
General convolution_backward function

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32708818).

jbschlosser committed:
Update on "General convolution_backward function"


Differential Revision: [D32708818](https://our.internmc.facebook.com/intern/diff/D32708818)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32708818).

jbschlosser committed:
Update on "General convolution_backward function"


Differential Revision: [D32708818](https://our.internmc.facebook.com/intern/diff/D32708818)

[ghstack-poisoned]

jbschlosser committed:
Update on "General convolution_backward function"


Differential Revision: [D32708818](https://our.internmc.facebook.com/intern/diff/D32708818)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32708818).

jbschlosser committed:
Update on "General convolution_backward function"


Differential Revision: [D32708818](https://our.internmc.facebook.com/intern/diff/D32708818)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32708818).

albanD commented on aten/src/ATen/native/Convolution.cpp:
```diff
@@ -985,6 +982,25 @@ static inline std::vector<int64_t> calc_output_size(
+    backend_memory_format = cudnn_conv_suggest_memory_format(input, weight);
+  }
+  if (detail::getCUDAHooks().compiledWithMIOpen() && miopen_conv_use_channels_last(input, weight)) {
+    backend_memory_format = (k == 5) ? at::MemoryFormat::Contiguous /*at::MemoryFormat::ChannelsLast3d*/ : at::MemoryFormat::ChannelsLast;
```
Is this assuming that k == 4 or 5 here?

jbschlosser replied:
Yep looks like it - it's technically valid at this point since the k == 3 case is explicitly expanded to 4D before this is called.

zou3519 replied:
can we add an internal assert on k == 4 or 5?

albanD reviewed (approved):
SGTM

jbschlosser committed:
Update on "General convolution_backward function"


Differential Revision: [D32708818](https://our.internmc.facebook.com/intern/diff/D32708818)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32708818).

NicolasHug commented:
![image](https://user-images.githubusercontent.com/1190450/144302698-c33625ab-d833-4884-9151-437b4c67deb0.png)


Looks like the ssh credential on the cluster are at it again?

jbschlosser committed:
Update on "General convolution_backward function"


Differential Revision: [D32708818](https://our.internmc.facebook.com/intern/diff/D32708818)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32708818).

jbschlosser committed:
Update on "General convolution_backward function"


Differential Revision: [D32708818](https://our.internmc.facebook.com/intern/diff/D32708818)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32708818).

zou3519 commented on test/test_nn.py:
```diff
@@ -13301,47 +13301,50 @@ def test_conv3d_valid_padding_backward(self, device):
+                decorators=[onlyNativeDeviceTypes, disableMkldnn, disablecuDNN], name='empty_channel3d'),
         subtest(((0, 0, 7, 8, 9), False, False, 3, torch.strided, torch._C._ConvBackend.Empty),
-                decorators=[onlyCPU, disableMkldnn], name='empty_batch_channel3d'),
+                decorators=[onlyNativeDeviceTypes, disableMkldnn, disablecuDNN], name='empty_batch_channel3d'),
```
I'm confused, why didn't we disable cudnn before?

jbschlosser replied:
Good question - I wrongly restricted the slow tests to CPU-only before. Now they run on CUDA as well (`onlyCPU` -> `onlyNativeDeviceTypes`), which means `disablecuDNN` should be set to ensure the slow algos are selected.

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32708818).

jbschlosser commented on aten/src/ATen/native/Convolution.cpp:
```diff
@@ -1483,4 +1488,289 @@ std::tuple<Tensor,Tensor,Tensor> _convolution_double_backward( const c10::option
+    case ConvBackend::Slow2d:
+      return at::_slow_conv2d_backward(
+        grad_output, input, weight, kernel_size, params.stride, params.padding, output_mask);
+    // NB: nnpack backward does not support strided convolutions; use slow impl instead
+    case ConvBackend::NnpackSpatial:
```
Yes that's correct - [it only supports 4d inputs](https://github.com/pytorch/pytorch/blob/b6f41bb8488e7a15705d015944cbfe2a912b6213/aten/src/ATen/native/NNPACK.cpp#L158-L162). The [old code](https://github.com/pytorch/pytorch/blob/b6f41bb8488e7a15705d015944cbfe2a912b6213/tools/autograd/derivatives.yaml#L2253-L2255) also utilized SlowDilated2d for the backward pass.

jbschlosser committed:
Update on "General convolution_backward function"


Differential Revision: [D32708818](https://our.internmc.facebook.com/intern/diff/D32708818)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32708818).

jbschlosser committed:
Update on "General convolution_backward function"


Differential Revision: [D32708818](https://our.internmc.facebook.com/intern/diff/D32708818)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32708818).

albanD reviewed (approved):
LGTM

jbschlosser committed:
Update on "General convolution_backward function"


Differential Revision: [D32708818](https://our.internmc.facebook.com/intern/diff/D32708818)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32708818).

albanD reviewed (approved):
👍

jbschlosser committed:
Update on "General convolution_backward function"


Differential Revision: [D32708818](https://our.internmc.facebook.com/intern/diff/D32708818)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32708818).

jbschlosser committed:
Update on "General convolution_backward function"


Differential Revision: [D32708818](https://our.internmc.facebook.com/intern/diff/D32708818)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32708818).

jbschlosser committed:
Update on "General convolution_backward function"


Differential Revision: [D32708818](https://our.internmc.facebook.com/intern/diff/D32708818)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32708818).

jbschlosser committed:
Update on "General convolution_backward function"


Differential Revision: [D32708818](https://our.internmc.facebook.com/intern/diff/D32708818)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32708818).

albanD commented on aten/src/ATen/native/ConvUtils.h:
```diff
@@ -62,10 +63,14 @@ enum class ConvBackend {
 
 // Function to select the convolution backend based on the inputs and params.
 // This overload is used within the convolution internals but not exposed to python.
+// NB: The forward pass provides a bias tensor while the backward pass provides
```
Comment needs to be updated right?

albanD commented on aten/src/ATen/native/ConvUtils.h:
```diff
@@ -62,10 +63,14 @@ enum class ConvBackend {
     const Tensor& weight,
-    const Tensor& bias,
+    const c10::optional<IntArrayRef> bias_sizes_opt,
+    const bool need_backward,
```
This would need a comment explaining what it means is?
I guess it is: grad mode is enabled and either input, weight or bias requires grad. ?

Also what about when this is called during backward? This means double backward will be needed?

albanD reviewed (approved):
Sounds good!

jbschlosser committed:
Update on "General convolution_backward function"


Differential Revision: [D32708818](https://our.internmc.facebook.com/intern/diff/D32708818)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32708818).
